### PR TITLE
Implement and provide default mappings for some Obsidian-specific Vim motions/commands

### PR DESCRIPTION
**Pull request by alythobani**
See discussion in [improved-obsidian-vimcursor](https://github.com/hhhapz/improved-obsidian-vimcursor/issues/1), which is no longer functional as a plugin due to Obsidian switching from CodeMirror 5 to 6.

## New motions/actions

This PR now has this plugin automatically provide the following vim motions/actions in Obsidian:
* **Jump to next/previous heading**: `[[` / `]]`
* **Jump to next/previous link/wikilink**: `gl` / `gL`
* **Move up/down, skipping folds**: `zk` / `zj`
* **Follow link under cursor, temporarily moving cursor if necessary**: `gf`

This PR also implements associated helper functions and types (`defineAndMapObsidianVimMotion`, `defineAndMapObsidianVimAction`, `jumpToPattern`, `MotionFn`, `ObsidianActionFn`) that would hopefully make it easier to add more motions/actions like these to the plugin in the future.

## Not done yet

* Update README and JSSnippets.md accordingly (want to get @esm7 's feedback here first before doing that)
* Unit tests (could write some if desired)
  * Testing manually, the commands do seem to work well, including on edge cases e.g. cursor already on a match, and various `count`s to jump multiple headings/links at a time)

## Screen recording

(Keystroke viewer may be a little hard/fast to see here, but including anyway in case it's useful)

https://github.com/esm7/obsidian-vimrc-support/assets/13757215/dcd292fb-86f5-4d6d-b82b-57f3f3444681

### CODE REVIEW

File: utils/jumpToPattern.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
+import { Editor as CodeMirrorEditor } from "codemirror";
+import { EditorPosition } from "obsidian";
+import { shim as matchAllShim } from "string.prototype.matchall";
+
+// Polyfill for String.prototype.matchAll, in case it's not available (pre-ES2020)
+matchAllShim();
```
Comment (alythobani) [inline utils/jumpToPattern.ts:6]: I'm guessing we don't actually need this polyfill since users are probably using a version of Obsidian that runs on a post-2020 Electron/Chromium version

**Comment by esm7** [comment]
First, I must express my deepest regrets for taking so long to review this. I think this is a *great* contribution to this plugin in so many ways, it's a shame users should have waited.
I made a few tiny updates to the branch which were required for me to build it and removed the polyfills. Other than that things seem to work well.
If you just update the plugin documentation, I think I can get this released this week.

**Comment by alythobani** [comment]
Hey @esm7  sorry for the slightly late reply myself and no worries at all, I know very well how life can get in the way of things like this :) thanks for getting around to looking through this, glad things look good to you on the whole!

I'll try and find time later today to update the documentation!

**Comment by alythobani** [comment]
Okay, just pushed some simple docs updates! @esm7 

Things could maybe be more organized since now I guess we have some outdated code examples in `JsSnippets.md`, but I figured maybe this is good enough for now, and those examples can still mainly serve as demonstrations for using JS commands for custom motions. But let me know if you'd like me to make any more changes.

**Comment by lebigot** [comment]
I tried this through the 0.10.0 plugin version, and the `]]` command also jumps to _comments_ inside ``` blocks, like in:

```
```tasks
# …
```

I'm not sure if this can easily be avoided, but I thought it doesn't hurt to mention this. :)

**Comment by alythobani** [comment]
Hey @lebigot glad you find it useful, and yeah thanks for pointing this out, I didn't think of this case! Someone else also noticed it and filed an issue #231  for it. I'm quite busy these next couple days, but this should definitely be fixable when I get a chance soon :) hopefully it's not the biggest annoyance in the meantime haha
